index.js

```diff
@@ -1,206 +1,229 @@
-var assign = Object.assign || require('es6-object-assign').assign
-var memoize = require('memoizee')
+'use strict'
 
-var parseArgForType = function (type, arg) {
+const memoize = require('memoizee')
+
+const defaultArgs = {
+  memops: {
+    maxAge: 10 * 60 * 1000,
+    /* 10 minutes */
+    preFetch: 0.5,
+    /* recache every 5 minutes */
+    primitive: true,
+    promise: true
+  }
+}
+
+const defaultConfig = {
+  args: {},
+  filters: {}
+}
+
+class Provider {
+
+  constructor(args = defaultArgs, config = defaultConfig) {
+    this.args = args
+    this.config = config
+
+    const memopts = args.memops
+
+    this.args = Object.assign(
+      this.args,
+      this._processArgs(this.args)
+    )
+    this.filters = Object.assign(
+      {},
+      Provider.DefaultFilters,
+      this.config.filters
+    )
+
+    this.fetch = this._makeCached(this.fetch, memopts)
+    this.detail = this._makeCached(this.detail, memopts)
+  }
+
+  _makeCached(method, memopts) {
+    const self = this
+    const memoizedMethod = memoize(method, memopts)
+
+    return function() {
+      // XXX: Should be replaced with spread operator if possible.
+      return memoizedMethod.apply(this, arguments)
```

Comment:
Arrow functions mess with the lexical scope: https://babeljs.io/learn-es2015/#ecmascript-2015-features-arrows-and-lexical-this